### PR TITLE
IPMPROG-5508 Added token document type

### PR DESCRIPTION
**Pull request by sinanyavas-eaton**
- Added Token and login document type
- Related unit tests implemented

configuration.json file might need to be changed depending on the use cases of the document.  

Signed-off-by: Sinan Yavas SinanYavas@eaton.com

### CODE REVIEW

**Comment by FrancoisRegisDegott-eaton** [comment]
do not merge in 2.7 release branch...

**Review by ericclappier-eaton — APPROVED** [review]

File: lib/include/secw_token_and_login.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+static constexpr const char* DOC_TOKEN_AND_LOGIN_TOKEN = "secw_token_and_login_token";
+static constexpr const char* DOC_TOKEN_AND_LOGIN_LOGIN = "secw_token_and_login_login";
+
+
+class TokenAndLogin final : public Document
+{
+public:
+    TokenAndLogin();
+
+    TokenAndLogin(const std::string& name, const std::string& login = "", const std::string& token = "");
```
Comment (FrancoisRegisDegott-eaton) [inline lib/include/secw_token_and_login.h]: Misleading.
follow the TokenAndLogin classname semantic, invert login/token args in ctor

File: lib/include/secw_token_and_login.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /// try to cast a document to a TokenAndLogin shared ptr
+    /// @return shared ptr on TokenAndLogin or null shared ptr in case of error
+    static TokenAndLoginPtr tryToCast(DocumentPtr doc);
+
+private:
+    // Public secw elements
+    std::string m_login;
+
```
Comment (FrancoisRegisDegott-eaton) [inline lib/include/secw_token_and_login.h]: \n  -1

File: lib/src/secw_token_and_login.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
+/*  =========================================================================
+    secw_login_and_token - Document parsers for TokenAndLogin document
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: secw_token_and_login

File: lib/src/secw_token_and_login.cc [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+/*-----------------------------------------------------------------------------*/
+/*   TokenAndLogin Document                                                              */
+/*-----------------------------------------------------------------------------*/
+// Public
+TokenAndLogin::TokenAndLogin()
+    : Document(TOKEN_AND_LOGIN_TYPE)
+{
+}
+
+TokenAndLogin::TokenAndLogin(const std::string& name, const std::string& login, const std::string& token)
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: misleading related to the semantic
invert login/token args order in ctor

File: lib/src/secw_token_and_login.cc [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    : Document(TOKEN_AND_LOGIN_TYPE)
+    , m_login(login)
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: same related to semantic

File: lib/src/secw_token_and_login.cc [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+DocumentPtr TokenAndLogin::clone() const
+{
+    return std::dynamic_pointer_cast<Document>(std::make_shared<TokenAndLogin>(*this));
+}
+
+void TokenAndLogin::validate() const
+{
+    if (!m_containPrivateData)
+        throw SecwInvalidDocumentFormatException(DOC_TOKEN_AND_LOGIN_TOKEN);
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc:60]: i don't understand the logic behind this "if (!containPrivateData) then throw()".
Any hints?

File: lib/src/secw_token_and_login.cc [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+void TokenAndLogin::fillSerializationInfoPrivateDoc(cxxtools::SerializationInfo& si) const
+{  
+    if (!m_token.empty()) {
+        si.addMember(DOC_TOKEN_AND_LOGIN_TOKEN) <<= StdStringToCxxString(m_token);
+    }
+}
+
+void TokenAndLogin::fillSerializationInfoPublicDoc(cxxtools::SerializationInfo& si) const
+{
+    if(!m_login.empty()) {
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: space missed

File: lib/src/secw_token_and_login.cc [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+void TokenAndLogin::updatePrivateDocFromSerializationInfo(const cxxtools::SerializationInfo& si)
+{
+    try {
+        const cxxtools::SerializationInfo* token = si.findMember(DOC_TOKEN_AND_LOGIN_TOKEN);
+        if (token != nullptr) {
+            m_token = GetSiMemberCxxString(si, DOC_TOKEN_AND_LOGIN_TOKEN);
+        }
+    } catch (const std::exception& e) {
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: e unused

File: lib/src/secw_token_and_login.cc [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+void TokenAndLogin::updatePublicDocFromSerializationInfo(const cxxtools::SerializationInfo& si)
+{
+    try {
+        const cxxtools::SerializationInfo* login = si.findMember(DOC_TOKEN_AND_LOGIN_LOGIN);
+        if (login != nullptr) {
+            m_login = GetSiMemberCxxString(si, DOC_TOKEN_AND_LOGIN_LOGIN);
+        }
+    } catch (const std::exception& e) {
```
Comment (FrancoisRegisDegott-eaton) [inline lib/src/secw_token_and_login.cc]: e unused

File: lib/tests/producer_accessor.cpp [hunk 10]
```diff
@@ -1522,4 +1522,245 @@ void secwProducerAccessorTest(fty::SocketSyncClient& syncClient, mlm::MlmStreamC
             FAIL(e.what());
         }
     }
+
+    // test 11.1 => insertNewDocument Login and Token
+    {
+        secw::ProducerAccessor producerAccessor(syncClient, streamClient);
+        try {
+            secw::TokenAndLoginPtr doc = std::make_shared<secw::TokenAndLogin>("insert test login token", "Login", "Token");
```
Comment (FrancoisRegisDegott-eaton) [inline lib/tests/producer_accessor.cpp]: args login/token order

File: lib/tests/producer_accessor.cpp [hunk 11]
```diff
@@ -1522,4 +1522,245 @@ void secwProducerAccessorTest(fty::SocketSyncClient& syncClient, mlm::MlmStreamC
...
+        } catch (const std::exception& e) {
+            FAIL(e.what());
+        }
+    }
+
+    // test 11.3 => insertNewDocument Login and Token => SecwNameAlreadyExistsException
+    {
+        secw::ProducerAccessor producerAccessor(syncClient, streamClient);
+        try {
+            secw::TokenAndLoginPtr doc = std::make_shared<secw::TokenAndLogin>("insert test login token", "Login", "Token");
```
Comment (FrancoisRegisDegott-eaton) [inline lib/tests/producer_accessor.cpp]: login/token args order

File: resources/configuration.json [hunk 12]
```diff
@@ -8,19 +8,22 @@
 					"Snmpv3",
 					"Snmpv1",
 					"UserAndPassword",
-					"ExternalCertificate"
+					"ExternalCertificate",
+					"TokenAndLogin"
```
Comment (FrancoisRegisDegott-eaton) [inline resources/configuration.json]: insert TokenAndLogin just after userAndpassword...
(ExternalCertificate is a 'special' document type)

File: lib/include/secw_token_and_login.h [hunk 13]
```diff
@@ -0,0 +1,88 @@
+/*  =========================================================================
+    secw_login_and_token - Document parsers for TokenAndLogin document
```
Comment (FrancoisRegisDegott-eaton) [inline lib/include/secw_token_and_login.h]: secw_token_and_login

File: lib/src/secw_token_and_login.cc (hunk 6 above)
Comment (sinanyavas-eaton) [reply lib/src/secw_token_and_login.cc:60]: it's for the validation of the document. docs without token is not allowed and should be caught at secw_security_wallet_server.cc:134.

**Review by FrancoisRegisDegott-eaton — APPROVED** [review]

**Pull request merged**